File: RegularGrammar/FiniteAutomaton.py
class FiniteAutomaton:
    def __init__(self):
        self.states = {'S', 'L', 'D', 'end'}
        self.alphabet = {'a', 'b', 'c', 'd', 'e', 'f', 'j'}
        self.transitions = {
            ('S', 'a'): {"S"},
            ('S', 'b'): {"S"},
            ('S', 'c'): {"D"},
            ('S', 'd'): {"L"},
            ('D', 'e'): {"D"},
            ('D', 'd'): {"end"},
            ('L', 'e'): {"L", "end"},
            ('L', 'f'): {"L"},
            ('L', 'j'): {"D"},
            ('S', 'e'): {"end"},
        }
        self.start_state = 'S'
        self.accept_states = {'end'}

    def accepts(self, input_string):
        current_states = {self.start_state}
        for char in input_string:
            next_states = set()
            for state in current_states:
                transition_key = (state, char)
                if transition_key in self.transitions:
                    next_states.update(self.transitions[transition_key])
            current_states = next_states

        return bool(current_states & self.accept_states)

File: FiniteAutomata/FiniteAutomaton.js
export class FiniteAutomaton {
  constructor() {
    this.states = ["q0", "q1", "q2", "q3"];
    this.alphabet = ["a", "b"];
    this.transitions = [
      { src: "q0", char: "a", dest: "q1" },
      { src: "q0", char: "a", dest: "q2" },
      { src: "q1", char: "b", dest: "q1" },
      { src: "q1", char: "a", dest: "q2" },
      { src: "q2", char: "a", dest: "q1" },
      { src: "q2", char: "b", dest: "q3" },
    ];
    this.start_state = "q0";
    this.accept_state = "q3";
  }

  to_regular_grammar() {
    let grammar = {
      terminals: [],
      non_terminals: [],
      rules: [],
      start: "",
    };

    this.alphabet.forEach((char) => {
      grammar.terminals.push(char);
    });

    this.states.forEach((state) => {
      grammar.non_terminals.push(state);
    });

    this.transitions.forEach((state) => {
      grammar.rules.push(state.src + "-" + state.char + "->" + state.dest);
    });

    grammar.start = this.start_state;

    return grammar;
  }
}
